// chain-sdk/dist/client.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Connection = require('./connection');
var accessTokensAPI = require('./api/accessTokens');
var accountsAPI = require('./api/accounts');
var assetsAPI = require('./api/assets');
var balancesAPI = require('./api/balances');
var configAPI = require('./api/config');
var mockHsmKeysAPI = require('./api/mockHsmKeys');
var transactionsAPI = require('./api/transactions');
var transactionFeedsAPI = require('./api/transactionFeeds');
var unspentOutputsAPI = require('./api/unspentOutputs');

/**
 * The Chain API Client object is the root object for all API interactions.
 * To interact with Chain Core, a Client object must always be instantiated
 * first.
 * @class
 */

var Client = function () {
  /**
   * constructor - create a new Chain client object capable of interacting with
   * the specified Chain Core.
   *
   * @param {String} baseUrl - Chain Core URL.
   * @param {String} token - Chain Core client token for API access.
   * @returns {Client}
   */
  function Client(baseUrl, token) {
    _classCallCheck(this, Client);

    this.connection = new Connection(baseUrl, token);

    /**
     * API actions for access tokens
     * @type {module:AccessTokensApi}
     */
    this.accessTokens = accessTokensAPI(this);

    /**
     * API actions for accounts
     * @type {module:AccountsApi}
     */
    this.accounts = accountsAPI(this);

    /**
     * API actions for assets.
     * @type {module:AssetsApi}
     */
    this.assets = assetsAPI(this);

    /**
     * API actions for balances.
     * @type {module:BalancesApi}
     */
    this.balances = balancesAPI(this);

    /**
     * API actions for config.
     * @type {module:ConfigApi}
     */
    this.config = configAPI(this);

    /**
     * @property {module:MockHsmKeysApi} keys API actions for Mock HSM keys.
     * @property {Connection} signerConnection Mock HSM signer connection.
     */
    this.mockHsm = {
      keys: mockHsmKeysAPI(this),
      signerConnection: new Connection('http://localhost:1999/mockhsm')
    };

    /**
     * API actions for transactions.
     * @type {module:TransactionsApi}
     */
    this.transactions = transactionsAPI(this);

    /**
     * API actions for transaction feeds.
     * @type {module:TransactionFeedsApi}
     */
    this.transactionFeeds = transactionFeedsAPI(this);

    /**
     * API actions for unspent outputs.
     * @type {module:UnspentOutputsApi}
     */
    this.unspentOutputs = unspentOutputsAPI(this);
  }

  /**
   * Submit a request to the stored Chain Core connection.
   *
   * @param {String} path
   * @param {object} [body={}]
   * @returns {Promise}
   */


  _createClass(Client, [{
    key: 'request',
    value: function request(path) {
      var body = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      return this.connection.request(path, body);
    }
  }]);

  return Client;
}();

module.exports = Client;
